File: app/src/test/java/org/mifos/mobile/LoginPresenterTest.java
package org.mifos.mobile;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobile.api.BaseURL;
import org.mifos.mobile.api.DataManager;
import org.mifos.mobile.api.local.PreferencesHelper;
import org.mifos.mobile.models.Page;
import org.mifos.mobile.models.User;
import org.mifos.mobile.models.client.Client;
import org.mifos.mobile.presenters.LoginPresenter;
import org.mifos.mobile.ui.views.LoginView;
import org.mifos.mobile.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;


import io.reactivex.Observable;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoginPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    PreferencesHelper mockHelper;

    @Mock
    LoginView view;

    private LoginPresenter presenter;
    private User user;
    private Page<Client> clientPage, noClientPage;

    @Before
    public void setUp() throws Exception {

        when(mockHelper.getBaseUrl()).thenReturn(BaseURL.PROTOCOL_HTTPS + BaseURL.API_ENDPOINT);
        when(dataManager.getPreferencesHelper()).thenReturn(mockHelper);

        presenter = new LoginPresenter(dataManager, context);
        presenter.attachView(view);

        user = FakeRemoteDataSource.getUser();
        clientPage = FakeRemoteDataSource.getClients();
        noClientPage = FakeRemoteDataSource.getNoClients();
    }

    @Test
    public void testLogin() throws Exception {
        when(dataManager.login("selfservice", "password")).thenReturn(Observable.just(user));

        presenter.login("selfservice", "password");

        verify(view).showProgress();
        verify(view).onLoginSuccess(user.getUsername());
    }

    @Test
    public void testLoadClients() throws Exception {
        long clientId = clientPage.getPageItems().get(0).getId();
        when(dataManager.getClients()).thenReturn(Observable.just(clientPage));

        presenter.loadClient();

        verify(view).hideProgress();
        verify(view).showPassCodeActivity();
        verify(view, never()).showMessage(context.getString(R.string.error_fetching_client));
    }

    @Test
    public void testLoadNoClients() throws Exception {
        long clientId = clientPage.getPageItems().get(0).getId();
        when(dataManager.getClients()).thenReturn(Observable.just(noClientPage));

        presenter.loadClient();

        verify(view).hideProgress();
        verify(view).showMessage(context.getString(R.string.error_client_not_found));
        verify(view, never()).showPassCodeActivity();
    }

    @Test
    public void testLoadClientFails() throws Exception {
        when(dataManager.getClients()).thenReturn(Observable.<Page<Client>>error(RetrofitUtils.
                get404Exception()));

        presenter.loadClient();

        verify(view).hideProgress();
        verify(view).showMessage(context.getString(R.string.error_fetching_client));
        verify(view, never()).showPassCodeActivity();
    }

    @Test
    public void testLoadClientUnauthorized() throws Exception {
        when(dataManager.getClients()).thenReturn(Observable.<Page<Client>>error(RetrofitUtils.
                get401Exception()));

        presenter.loadClient();

        verify(view).hideProgress();
        verify(view).showMessage(context.getString(R.string.unauthorized_client));
        verify(view, never()).showPassCodeActivity();
    }
    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }
}

File: app/src/test/java/org/mifos/mobile/LoginPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.BaseURL
import org.mifos.mobile.api.DataManager
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.models.Page
import org.mifos.mobile.models.User
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.models.payload.LoginPayload
import org.mifos.mobile.presenters.LoginPresenter
import org.mifos.mobile.ui.views.LoginView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner::class)
class LoginPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var mockHelper: PreferencesHelper? = null

    @Mock
    var view: LoginView? = null
    private var presenter: LoginPresenter? = null
    private var user: User? = null
    private var clientPage: Page<Client?>? = null
    private var noClientPage: Page<Client?>? = null
    private var loginPayload: LoginPayload? = null
    private var emptyClientName: String = ""

    @Before
    @Throws(Exception::class)
    fun setUp() {
        Mockito.`when`(mockHelper?.baseUrl).thenReturn(BaseURL.PROTOCOL_HTTPS + BaseURL.API_ENDPOINT)
        Mockito.`when`(dataManager?.preferencesHelper).thenReturn(mockHelper)
        presenter = LoginPresenter(dataManager, context)
        presenter?.attachView(view)
        user = FakeRemoteDataSource.user
        clientPage = FakeRemoteDataSource.clients
        noClientPage = FakeRemoteDataSource.noClients
        loginPayload = FakeRemoteDataSource.loginPayload
    }

    @Test
    @Throws(Exception::class)
    fun testLogin() {
        Mockito.`when`(dataManager?.login(loginPayload)).thenReturn(Observable.just(user))
        presenter?.login(loginPayload)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.onLoginSuccess()
    }

    @Test
    @Throws(Exception::class)
    fun testLoadClients() {
        val clientName: String? = clientPage?.pageItems?.get(0)?.displayName
        Mockito.`when`(dataManager?.clients).thenReturn(Observable.just(clientPage))
        presenter?.loadClient()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showPassCodeActivity(clientName)
        Mockito.verify(view, Mockito.never())?.showMessage(context?.getString(R.string.error_fetching_client))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadNoClients() {
        Mockito.`when`(dataManager?.clients).thenReturn(Observable.just(noClientPage))
        presenter?.loadClient()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showMessage(context?.getString(R.string.error_client_not_found))
        Mockito.verify(view, Mockito.never())?.showPassCodeActivity(emptyClientName)
    }

    @Test
    @Throws(Exception::class)
    fun testLoadClientFails() {
        Mockito.`when`(dataManager?.clients).thenReturn(Observable.error(RetrofitUtils.get404Exception()))
        presenter?.loadClient()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showMessage(context?.getString(R.string.error_fetching_client))
        Mockito.verify(view, Mockito.never())?.showPassCodeActivity(emptyClientName)
    }

    @Test
    @Throws(Exception::class)
    fun testLoadClientUnauthorized() {
        Mockito.`when`(dataManager?.clients).thenReturn(Observable.error(RetrofitUtils.get401Exception()))
        presenter?.loadClient()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showMessage(context?.getString(R.string.unauthorized_client))
        Mockito.verify(view, Mockito.never())?.showPassCodeActivity(emptyClientName)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }
}